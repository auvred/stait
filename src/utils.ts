export type ContextDecl = Record<string, unknown>

export type ActionsDecl<_Typestait extends { type: string }> = Record<
  string,
  // `any` looks cleaner in user-facing errors
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (payload?: any) => _Typestait['type']
>

export type Flatten<T> = [T] extends [infer TT]
  ? {
      [K in keyof TT]: TT[K]
    }
  : never
