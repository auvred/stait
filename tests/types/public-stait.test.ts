import { runTypesTests } from '../utils'

import type {
  PublicActionFromTypestait,
  PublicStaitFromTypestait,
  Typestait,
} from '../../src'
import type { Equal } from '../utils'

runTypesTests(import.meta.url)

type ExpectType_public_stait = PublicStaitFromTypestait<
  Typestait<{
    type: 'stait1'
    context: {
      data1: string
      data2: 123
    }
    actions: {
      action1: (payload: { data3: string; data4: '123' }) => 'stait1'
    }
  }>,
  'stait1'
>

type ExpectError_should_validate_requested_type = PublicStaitFromTypestait<
  Typestait<{
    type: 'stait1'
    context: {
      data1: string
      data2: 123
    }
    actions: {
      action1: (payload: { data3: string; data4: '123' }) => 'stait1'
    }
  }>,
  'stait3'
>

type ExpectType_should_hide_service = Equal<
  PublicStaitFromTypestait<
    Typestait<{
      type: 'stait1'
      context: {
        data1: string
        data2: 123
      }
      service: 'stait1'
    }>,
    'stait1'
  >,
  {
    type: 'stait1'
    context: {
      data1: string
      data2: 123
    }
  }
>

type ExpectType_index_accessed_action = PublicStaitFromTypestait<
  Typestait<{
    type: 'stait1'
    actions: {
      action1: (payload: { data1: string; data2: 123 }) => 'stait1'
    }
  }>,
  'stait1'
>['actions']['action1']
type ExpectType_public_action_from_typestait = PublicActionFromTypestait<
  Typestait<{
    type: 'stait1'
    actions: {
      action1: (payload: { data1: string; data2: 123 }) => 'stait1'
    }
  }>,
  'stait1',
  'action1'
>
type ExpectType_index_accessed_action_equals_public_action_from_typestait =
  Equal<
    ExpectType_index_accessed_action,
    ExpectType_public_action_from_typestait
  >

type ExpectError_public_action_from_typestait_should_validate_type =
  PublicActionFromTypestait<
    Typestait<{
      type: 'stait1'
      actions: {
        action1: () => 'stait1'
      }
    }>,
    'stait2',
    'action1'
  >

type ExpectError_public_action_from_typestait_should_validate_action_name =
  PublicActionFromTypestait<
    Typestait<{
      type: 'stait1'
      actions: {
        action1: () => 'stait1'
      }
    }>,
    'stait1',
    'action2'
  >

type ExpectType_action_without_payload = PublicActionFromTypestait<
  Typestait<{
    type: 'stait1'
    actions: {
      action1: () => 'stait1'
    }
  }>,
  'stait1',
  'action1'
>
type ExpectType_action_should_preserve_absence_of_payload = Equal<
  Parameters<ExpectType_action_without_payload>,
  []
>

type ExpectType_action_with_payload = PublicActionFromTypestait<
  Typestait<{
    type: 'stait1'
    actions: {
      action1: (payload: { data1: string; data2: 123 }) => 'stait1'
    }
  }>,
  'stait1',
  'action1'
>
type ExpectType_action_should_preserve_presence_of_payload = Equal<
  Parameters<ExpectType_action_with_payload>,
  [{ data1: string; data2: 123 }]
>
