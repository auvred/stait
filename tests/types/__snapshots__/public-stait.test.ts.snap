// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`error: public action from typestait should validate action name 1`] = `
"error TS2344: Type '\\"action2\\"' does not satisfy the constraint '\\"action1\\"'.
"
`;

exports[`error: public action from typestait should validate type 1`] = `
"error TS2344: Type '\\"stait2\\"' does not satisfy the constraint '\\"stait1\\"'.
"
`;

exports[`error: should validate requested type 1`] = `
"error TS2344: Type '\\"stait3\\"' does not satisfy the constraint '\\"stait1\\"'.
"
`;

exports[`type: index accessed action 1`] = `
"type index_accessed_action =
  (payload: {
    data1: string
    data2: 123
  }) => void
"
`;

exports[`type: index accessed action equals public action from typestait 1`] = `
"type index_accessed_action_equals_public_action_from_typestait =
  true
"
`;

exports[`type: public action from typestait 1`] = `
"type public_action_from_typestait =
  (payload: {
    data1: string
    data2: 123
  }) => void
"
`;

exports[`type: public stait 1`] = `
"type public_stait =
  {
    type: 'stait1'
    context: {
      data1: string
      data2: 123
    }
    actions: {
      action1: (payload: {
        data3: string
        data4: '123'
      }) => void
    }
  }
"
`;

exports[`type: should hide service 1`] = `
"type should_hide_service =
  true
"
`;
